### PR TITLE
Adding the NLP task summarization

### PR DESCRIPTION
**Pull request by liususan091219**
Adding the nlp task summarization and seq2seq

### CODE REVIEW

File: flaml/ml.py [hunk 1]
```diff
@@ -155,9 +155,14 @@ def metric_loss_score(
 
                 metric = datasets.load_metric(metric_name)
                 metric_mode = huggingface_metric_to_mode[metric_name]
-                score = metric.compute(predictions=y_predict, references=y_true)[
-                    metric_name
-                ]
+
+                if metric_name == "rouge":
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/ml.py:159]: rouge1 is a sub metric for rouge. 

When loading rouge we need to use metric = datasets.load_metric("rouge"), then the score contains rouge1, rouge2:

FileNotFoundError: Couldn't find a metric script at /home/xliu127/rouge1/rouge1.py. Metric 'rouge1' doesn't exist on the Hugging Face Hub either.

File: flaml/automl.py [hunk 2]
```diff
@@ -920,7 +922,7 @@ def _validate_data(
                 self._state.X_val = self._transformer.transform(X_val)
             else:
                 self._state.X_val = X_val
-            if self._label_transformer:
+            if self._label_transformer and self._state.task not in NLG_TASKS:
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/automl.py:925]: If it's NLG_TASKS, y_val is a pandas series containing the output sequence tokens, so we cannot use label_transformer.transform to process it.

[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/automl.py:925]: Added to comments

File: flaml/nlp/utils.py [hunk 3]
```diff
@@ -32,61 +25,140 @@ def load_default_huggingface_metric_for_task(task):
...
-    )
-    d = X.apply(
-        lambda x: tokenize_glue(x, this_tokenizer, custom_hpo_args),
-        axis=1,
-        result_type="expand",
-    )
+    if this_tokenizer:
+        with this_tokenizer.as_target_tokenizer():
+            d = X.apply(
+                lambda x: tokenize_glue(
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/nlp/utils.py:77]: No, it works for any sentence and sentence pair.

File: flaml/nlp/utils.py [hunk 4]
```diff
@@ -32,61 +25,140 @@ def load_default_huggingface_metric_for_task(task):
...
     return [tokenized_example[x] for x in tokenized_column_names]
 
 
-def separate_config(config):
-    from transformers import TrainingArguments
+def separate_config(config, task):
+    if task in NLG_TASKS:
+        from transformers import Seq2SeqTrainingArguments, TrainingArguments
+
+        trainargs_class_list = [Seq2SeqTrainingArguments, TrainingArguments]
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/nlp/utils.py:148]: yes

File: flaml/nlp/huggingface/trainer.py [hunk 5]
```diff
@@ -25,19 +60,22 @@ def evaluate(
...
+                self,
+                eval_dataset,
+                ignore_keys,
+                metric_key_prefix,
+                max_length=self.args.generation_max_length,
+                num_beams=self.args.generation_num_beams,
+            )
+        else:
+            metrics = eval_dataset and TFTrainer.evaluate(
+                self,
```
Comment (sonichi) [inline flaml/nlp/huggingface/trainer.py:74]:
```suggestion
            metrics = eval_dataset and super(Seq2SeqTrainer, self).evaluate(
```

File: flaml/nlp/huggingface/trainer.py [hunk 6]
```diff
@@ -7,13 +7,48 @@
...
+    def prediction_step(
+        self,
+        model,
+        inputs,
+        prediction_loss_only,
+        ignore_keys,
+    ):
+        if hasattr(self, "_is_seq2seq") and self._is_seq2seq:
+            return Seq2SeqTrainer.prediction_step(
+                self, model, inputs, prediction_loss_only, ignore_keys
```
Comment (sonichi) [inline flaml/nlp/huggingface/trainer.py:40]:
```suggestion
            return super().prediction_step(
                model, inputs, prediction_loss_only, ignore_keys
```

File: flaml/nlp/huggingface/trainer.py [hunk 7]
```diff
@@ -7,13 +7,48 @@
...
+            )
+        else:
+            return TFTrainer.prediction_step(
+                self, model, inputs, prediction_loss_only, ignore_keys
```
Comment (sonichi) [inline flaml/nlp/huggingface/trainer.py:44]:
```suggestion
            return super(Seq2SeqTrainer, self).prediction_step(
                model, inputs, prediction_loss_only, ignore_keys
```

File: flaml/nlp/huggingface/trainer.py [hunk 8]
```diff
@@ -7,13 +7,49 @@
     TFTrainer = object
```
Comment (sonichi) [inline flaml/nlp/huggingface/trainer.py:7]:
```suggestion
    Seq2SeqTrainer = object
```
Remove import of Trainer

File: flaml/automl.py [hunk 9]
```diff
@@ -1431,6 +1439,9 @@ def _decide_split_type(self, split_type):
             ), "groups must be specified for ranking task."
             assert split_type in ["auto", "group"]
             self._split_type = "group"
+        elif self._state.task == SUMMARIZATION:
```
Comment (sonichi) [inline flaml/automl.py:1442]: For the support of new tasks, several places need to be updated:
1. docstr about supported task types.
2. documentation: Use Cases/Task-Oriented AutoML; Examples/AutoML for NLP.
3. notebook
If you plan to add some of these in the next PR, please create some TODO tasks in project.

Comment (sonichi) [reply flaml/automl.py:1442]: What about NLG_TASKS?

Comment (liususan091219) [reply flaml/automl.py:1442]: "What about NLG_TASKS?": resolved

Comment (liususan091219) [reply flaml/automl.py:1442]: 1, 2: resolved

3: where to create TODOs? I don't plan on notebook in this PR

File: flaml/nlp/huggingface/trainer.py [hunk 10]
```diff
@@ -4,7 +4,7 @@
     from transformers import Trainer as TFTrainer
```
Comment (sonichi) [inline flaml/nlp/huggingface/trainer.py:4]: do we still need this import?

File: website/docs/Examples/AutoML-NLP.md [hunk 11]
```diff
@@ -86,4 +86,43 @@ automl_settings["custom_hpo_args"] = {
 automl.fit(
     X_train=X_train, y_train=y_train, X_val=X_val, y_val=y_val, **automl_settings
 )
+```
+
+### A simple summarization example
```
Comment (sonichi) [inline website/docs/Examples/AutoML-NLP.md:91]: Have you tested this example locally?
If possible, display some example output like in other examples.

File: flaml/nlp/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/nlp/utils.py:77]: Why is the function named tokenize_glue?

File: flaml/ml.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/ml.py:159]: what if the user wants to use "rouge2"?

Comment (sonichi) [inline flaml/ml.py:159]:
```suggestion
                if "rouge" in metric_name:
```

File: flaml/ml.py [hunk 12]
```diff
@@ -155,9 +155,14 @@ def metric_loss_score(
 
                 metric = datasets.load_metric(metric_name)
                 metric_mode = huggingface_metric_to_mode[metric_name]
-                score = metric.compute(predictions=y_predict, references=y_true)[
-                    metric_name
-                ]
+
+                if metric_name == "rouge":
+                    score = metric.compute(predictions=y_predict, references=y_true)["rouge1"].mid.fmeasure
```
Comment (sonichi) [inline flaml/ml.py:160]:
```suggestion
                    score = metric.compute(predictions=y_predict, references=y_true)[metric_name].mid.fmeasure

```

File: flaml/nlp/utils.py [hunk 13]
```diff
@@ -118,7 +118,7 @@ def postprocess_text(preds, labels):
     return preds, labels
 
 
-def tokenize_glue(
+def tokenize_each_example(
```
Comment (sonichi) [inline flaml/nlp/utils.py:121]:
```suggestion
def tokenize_row(
```

**Review by sonichi — APPROVED** [review]

**Pull request merged**